File: backend/internal/storage/personal_insights_test.go
package storage

import (
	"context"
	"math"
	"path/filepath"
	"testing"
)

func newPersonalInsightsTestStore(t *testing.T) *SQLiteStore {
	t.Helper()
	store, err := NewSQLiteStore(filepath.Join(t.TempDir(), "personal-insights.db"))
	if err != nil {
		t.Fatal(err)
	}
	if err := store.Migrate(context.Background()); err != nil {
		_ = store.Close()
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = store.Close() })
	return store
}

func seedPersonalInsightsMovie(t *testing.T, store *SQLiteStore, id, studio string, rating *float64) {
	t.Helper()
	if _, err := store.db.Exec(`
		INSERT INTO movies (
			id, title, code, studio, summary, added_at, location, resolution, year, user_rating
		) VALUES (?, ?, ?, ?, '', '2026-01-01', ?, '1080p', 2026, ?)`,
		id, id, id, studio, filepath.Join(t.TempDir(), id+".mp4"), rating,
	); err != nil {
		t.Fatal(err)
	}
}

func seedPersonalInsightsWatch(t *testing.T, store *SQLiteStore, day, movieID string, seconds float64) {
	t.Helper()
	if _, err := store.db.Exec(`
		INSERT INTO playback_daily_watch_time (day_key, movie_id, watched_sec, updated_at)
		VALUES (?, ?, ?, '2026-07-21T00:00:00Z')`, day, movieID, seconds); err != nil {
		t.Fatal(err)
	}
}

func TestPersonalInsightsAggregatesOverviewAndBoundedBreakdowns(t *testing.T) {
	store := newPersonalInsightsTestStore(t)
	ctx := context.Background()
	ratingA := 4.5
	ratingC := 3.5
	seedPersonalInsightsMovie(t, store, "movie-a", "Studio Alpha", &ratingA)
	seedPersonalInsightsMovie(t, store, "movie-b", "Studio Beta", nil)
	seedPersonalInsightsMovie(t, store, "movie-c", "Studio Alpha", &ratingC)
	seedPersonalInsightsWatch(t, store, "2026-07-01", "movie-a", 120)
	seedPersonalInsightsWatch(t, store, "2026-07-02", "movie-a", 30)
	seedPersonalInsightsWatch(t, store, "2026-07-02", "movie-b", 50)
	seedPersonalInsightsWatch(t, store, "2026-06-01", "movie-c", 80)
	if _, err := store.db.Exec(`
		INSERT INTO playback_progress (movie_id, position_sec, duration_sec, updated_at)
		VALUES ('movie-a', 90, 100, '2026-07-02T00:00:00Z'),
		       ('movie-b', 20, 100, '2026-07-02T00:00:00Z')`); err != nil {
		t.Fatal(err)
	}

	actorAlice := seedActorMergeActor(t, store, "Alice", actorMergeFields(nil))
	actorBob := seedActorMergeActor(t, store, "Bob", actorMergeFields(nil))
	actorShared := seedActorMergeActor(t, store, "Shared", actorMergeFields(nil))
	for _, pair := range []struct {
		movieID string
		actorID int64
	}{{"movie-a", actorAlice}, {"movie-a", actorShared}, {"movie-b", actorBob}, {"movie-b", actorShared}} {
		if _, err := store.db.Exec(`INSERT INTO movie_actors (movie_id, actor_id) VALUES (?, ?)`, pair.movieID, pair.actorID); err != nil {
			t.Fatal(err)
		}
	}
	for _, tag := range []struct {
		name    string
		typeKey string
		movies  []string
	}{
		{"Action", "nfo", []string{"movie-a", "movie-b"}},
		{"Action", "user", []string{"movie-a"}},
		{"Drama", "nfo", []string{"movie-a"}},
	} {
		result, err := store.db.Exec(`INSERT INTO tags (name, type) VALUES (?, ?)`, tag.name, tag.typeKey)
		if err != nil {
			t.Fatal(err)
		}
		tagID, _ := result.LastInsertId()
		for _, movieID := range tag.movies {
			if _, err := store.db.Exec(`INSERT INTO movie_tags (movie_id, tag_id) VALUES (?, ?)`, movieID, tagID); err != nil {
				t.Fatal(err)
			}
		}
	}

	overview, err := store.PersonalInsightsOverview(ctx, "2026-07-01", "2026-07-02", 0.90)
	if err != nil {
		t.Fatal(err)
	}
	if overview.WatchedSeconds != 200 || overview.StartedMovies != 2 ||
		overview.CompletedMovies != 1 || overview.RatedMovies != 1 ||
		overview.AverageRating == nil || *overview.AverageRating != 4.5 {
		t.Fatalf("unexpected overview: %+v", overview)
	}
	dataSince, err := store.PersonalInsightsDataSince(ctx, "2026-07-02")
	if err != nil || dataSince == nil || *dataSince != "2026-06-01" {
		t.Fatalf("unexpected dataSince=%v err=%v", dataSince, err)
	}

	assertBreakdown := func(dimension string, expected []PersonalInsightsBreakdownAggregate) {
		t.Helper()
		items, err := store.PersonalInsightsBreakdown(ctx, "2026-07-01", "2026-07-02", dimension, 10)
		if err != nil {
			t.Fatal(err)
		}
		if len(items) != len(expected) {
			t.Fatalf("%s item count=%d want=%d: %+v", dimension, len(items), len(expected), items)
		}
		for index, want := range expected {
			got := items[index]
			if got.Name != want.Name || got.MovieCount != want.MovieCount || math.Abs(got.WatchedSeconds-want.WatchedSeconds) > 0.001 {
				t.Fatalf("%s item[%d]=%+v want=%+v", dimension, index, got, want)
			}
		}
	}
	assertBreakdown("actor", []PersonalInsightsBreakdownAggregate{
		{Name: "Shared", WatchedSeconds: 200, MovieCount: 2},
		{Name: "Alice", WatchedSeconds: 150, MovieCount: 1},
		{Name: "Bob", WatchedSeconds: 50, MovieCount: 1},
	})
	assertBreakdown("studio", []PersonalInsightsBreakdownAggregate{
		{Name: "Studio Alpha", WatchedSeconds: 150, MovieCount: 1},
		{Name: "Studio Beta", WatchedSeconds: 50, MovieCount: 1},
	})
	assertBreakdown("tag", []PersonalInsightsBreakdownAggregate{
		{Name: "Action", WatchedSeconds: 200, MovieCount: 2},
		{Name: "Drama", WatchedSeconds: 150, MovieCount: 1},
	})
}

func TestPersonalInsightsOverviewReturnsNullDenominatorsForEmptyRange(t *testing.T) {
	store := newPersonalInsightsTestStore(t)
	overview, err := store.PersonalInsightsOverview(context.Background(), "2026-07-01", "2026-07-02", 0.90)
	if err != nil {
		t.Fatal(err)
	}
	if overview.WatchedSeconds != 0 || overview.StartedMovies != 0 || overview.AverageRating != nil {
		t.Fatalf("unexpected empty overview: %+v", overview)
	}
}

func TestPersonalInsightsBreakdownUsesMovieCountNameTieBreakAndLimit(t *testing.T) {
	store := newPersonalInsightsTestStore(t)
	for _, fixture := range []struct {
		id      string
		studio  string
		seconds float64
	}{
		{id: "two-a", studio: "Two", seconds: 30},
		{id: "two-b", studio: "Two", seconds: 30},
		{id: "alpha", studio: "alpha", seconds: 60},
		{id: "zeta", studio: "Zeta", seconds: 60},
	} {
		seedPersonalInsightsMovie(t, store, fixture.id, fixture.studio, nil)
		seedPersonalInsightsWatch(t, store, "2026-07-01", fixture.id, fixture.seconds)
	}

	items, err := store.PersonalInsightsBreakdown(
		context.Background(),
		"2026-07-01",
		"2026-07-01",
		"studio",
		2,
	)
	if err != nil {
		t.Fatal(err)
	}
	if len(items) != 2 ||
		items[0].Name != "Two" || items[0].MovieCount != 2 ||
		items[1].Name != "alpha" || items[1].MovieCount != 1 {
		t.Fatalf("unexpected bounded tie order: %+v", items)
	}
}

File: src/lib/personal-insights.test.ts
import { describe, expect, it } from "vitest"
import type { Movie } from "@/domain/movie/types"
import type { PlaybackProgressEntry } from "@/lib/playback-progress-storage"
import type { PlaybackWatchTimeMovieEntry } from "@/lib/playback-watch-time-storage"
import { buildPersonalInsightsBreakdown, buildPersonalInsightsOverview } from "./personal-insights"

const now = new Date("2026-07-21T16:30:00.000Z")

function movie(id: string, overrides: Partial<Movie> = {}): Movie {
  return {
    id,
    title: id,
    code: id.toUpperCase(),
    studio: "",
    actors: [],
    tags: [],
    userTags: [],
    runtimeMinutes: 120,
    rating: 4,
    summary: "",
    isFavorite: false,
    addedAt: "2026-01-01",
    location: `D:/media/${id}.mp4`,
    resolution: "1080p",
    year: 2026,
    tone: "",
    coverClass: "",
    ...overrides,
  }
}

const movies: Movie[] = [
  movie("movie-a", {
    studio: "Studio Alpha",
    actors: ["Alice", "Shared"],
    tags: ["Action", "Drama"],
    userTags: [" action "],
    userRating: 4.5,
  }),
  movie("movie-b", {
    studio: "Studio Beta",
    actors: ["Bob", "Shared"],
    tags: ["Action"],
  }),
  movie("movie-c", {
    studio: "Studio Alpha",
    actors: ["Carol"],
    tags: ["Classic"],
    userRating: 3.5,
  }),
]

const progressEntries: PlaybackProgressEntry[] = [
  { movieId: "movie-a", positionSec: 90, durationSec: 100, updatedAt: "2026-07-02T00:00:00Z" },
  { movieId: "movie-b", positionSec: 20, durationSec: 100, updatedAt: "2026-07-02T00:00:00Z" },
]

const watchTimeEntries: PlaybackWatchTimeMovieEntry[] = [
  { dayKey: "2026-07-01", movieId: "movie-a", watchedSec: 120 },
  { dayKey: "2026-07-02", movieId: "movie-a", watchedSec: 30 },
  { dayKey: "2026-07-02", movieId: "movie-b", watchedSec: 50 },
  { dayKey: "2026-06-01", movieId: "movie-c", watchedSec: 80 },
  { dayKey: "2026-07-01", movieId: "missing-movie", watchedSec: 999 },
  { dayKey: "2026-08-01", movieId: "movie-a", watchedSec: 999 },
]

describe("personal insights aggregates", () => {
  it("uses inclusive local calendar windows and explicit current-state denominators", () => {
    const overview = buildPersonalInsightsOverview({
      movies,
      progressEntries,
      watchTimeEntries,
      range: "30d",
      timezone: "Asia/Shanghai",
      now,
    })

    expect(overview).toEqual({
      range: "30d",
      from: "2026-06-23",
      to: "2026-07-22",
      timezone: "Asia/Shanghai",
      generatedAt: "2026-07-21T16:30:00.000Z",
      dataSince: "2026-06-01",
      watchedSeconds: 200,
      startedMovies: 2,
      completedMovies: 1,
      completionRate: 0.5,
      completionThreshold: 0.9,
      ratedMovies: 1,
      averageUserRating: 4.5,
    })
  })

  it("returns null rates for empty denominators instead of misleading zeroes", () => {
    const overview = buildPersonalInsightsOverview({
      movies,
      progressEntries: [],
      watchTimeEntries: [],
      range: "90d",
      timezone: "UTC",
      now,
    })

    expect(overview).toMatchObject({
      from: "2026-04-23",
      to: "2026-07-21",
      dataSince: null,
      watchedSeconds: 0,
      startedMovies: 0,
      completedMovies: 0,
      completionRate: null,
      ratedMovies: 0,
      averageUserRating: null,
    })
  })

  it("uses dataSince for all-time and includes every non-future library row", () => {
    const overview = buildPersonalInsightsOverview({
      movies,
      progressEntries,
      watchTimeEntries,
      range: "all",
      timezone: "UTC",
      now,
    })

    expect(overview).toMatchObject({
      from: "2026-06-01",
      to: "2026-07-21",
      watchedSeconds: 280,
      startedMovies: 3,
      ratedMovies: 2,
      averageUserRating: 4,
    })
  })

  it("applies full-per-entity attribution, per-movie tag dedupe, stable order, and limit", () => {
    const source = {
      movies,
      progressEntries,
      watchTimeEntries,
      range: "30d" as const,
      timezone: "Asia/Shanghai",
      now,
    }

    expect(buildPersonalInsightsBreakdown({ ...source, dimension: "actor" }).items).toEqual([
      { name: "Shared", watchedSeconds: 200, movieCount: 2, shareOfTotal: 1 },
      { name: "Alice", watchedSeconds: 150, movieCount: 1, shareOfTotal: 0.75 },
      { name: "Bob", watchedSeconds: 50, movieCount: 1, shareOfTotal: 0.25 },
    ])
    expect(buildPersonalInsightsBreakdown({ ...source, dimension: "studio" }).items).toEqual([
      { name: "Studio Alpha", watchedSeconds: 150, movieCount: 1, shareOfTotal: 0.75 },
      { name: "Studio Beta", watchedSeconds: 50, movieCount: 1, shareOfTotal: 0.25 },
    ])
    const tags = buildPersonalInsightsBreakdown({ ...source, dimension: "tag", limit: 1 })
    expect(tags).toMatchObject({
      totalWatchedSeconds: 200,
      attribution: "full-per-entity",
      limit: 1,
    })
    expect(tags.items).toEqual([
      { name: "Action", watchedSeconds: 200, movieCount: 2, shareOfTotal: 1 },
    ])
  })

  it("rejects invalid timezone and out-of-contract limits", () => {
    expect(() => buildPersonalInsightsOverview({
      movies,
      progressEntries,
      watchTimeEntries,
      range: "30d",
      timezone: "Mars/Olympus",
      now,
    })).toThrow("timezone is invalid")
    expect(() => buildPersonalInsightsBreakdown({
      movies,
      progressEntries,
      watchTimeEntries,
      range: "30d",
      timezone: "UTC",
      now,
      dimension: "tag",
      limit: 26,
    })).toThrow("between 1 and 25")
  })

  it("uses movie count then stable case-insensitive name ordering before applying the limit", () => {
    const tieMovies = [
      movie("two-a", { studio: "Two" }),
      movie("two-b", { studio: "Two" }),
      movie("alpha", { studio: "alpha" }),
      movie("zeta", { studio: "Zeta" }),
    ]
    const tieRows = [
      { dayKey: "2026-07-01", movieId: "two-a", watchedSec: 30 },
      { dayKey: "2026-07-01", movieId: "two-b", watchedSec: 30 },
      { dayKey: "2026-07-01", movieId: "alpha", watchedSec: 60 },
      { dayKey: "2026-07-01", movieId: "zeta", watchedSec: 60 },
    ]

    const result = buildPersonalInsightsBreakdown({
      movies: tieMovies,
      progressEntries: [],
      watchTimeEntries: tieRows,
      range: "30d",
      timezone: "UTC",
      now,
      dimension: "studio",
      limit: 2,
    })

    expect(result.items.map((item) => [item.name, item.movieCount])).toEqual([
      ["Two", 2],
      ["alpha", 1],
    ])
  })
})
